### PR TITLE
Add context menu for column headers for DLT files

### PR DESCRIPTION
**Pull request by sudeeptarlekar**
Chipmunk2 has an ability to colorise the columns as we as hide/show columns from DLT file.
Chipmunk3 is missing this feature and this PR adds back same feature.

Fixes: #1852

### CODE REVIEW

**Comment by sudeeptarlekar** [comment]
https://github.com/esrlabs/chipmunk/assets/23317691/52099de2-a2df-404a-9eb2-38455d6b77d5

File: application/client/src/app/ui/views/workspace/headers/menu/template.html [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+            <span
+                class="label t-normal"
+                [ngStyle]="{'color': header.color}">
+                {{header.caption}}
+            </span>
+        </li>
+    </ul>
+    <div class="controls">
+        <ul class="colors">
+            <li [attr.class]="'color'"
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/menu/template.html:20]:
```
[attr.class]="'color'"
```
no need to link it dynamically. Can be just 
```
class="color"
```

File: application/client/src/app/ui/views/workspace/headers/menu/template.html [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<div class="wrapper">
+    <ul class="columns">
+        <li
+            [attr.class]="'column'"
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/menu/template.html:4]: no need to link it dynamically. Can be just

```
[attr.class]="'column'"
```
can be
```
class="column"
```

File: application/client/src/app/ui/views/workspace/headers/menu/component.ts [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    templateUrl: './template.html',
+    changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class ViewWorkspaceHeadersMenuComponent extends ChangesDetector {
+    public selectedColumn: number | undefined = undefined;
+    public colors: string[] = CColors;
+
+    @Input() public column!: number;
+    @Input() public controller!: Columns;
+    @Input() public header!: string;
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/menu/component.ts:24]: `header` is this used somewhere?

File: application/client/src/app/ui/views/workspace/headers/component.ts [hunk 4]
```diff
@@ -63,13 +69,44 @@ export class ColumnsHeaders extends ChangesDetector implements AfterContentInit
...
+    }
+
+    public ngOnClick(event: MouseEvent, column: number): void {
+        contextmenu.show({
+            component: {
+                factory: ViewWorkspaceHeadersMenuComponent,
+                inputs: {
+                    column,
+                    controller: this.controller,
+                    header: this.ngMouseOverHeader,
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/component.ts:102]: `header` is this used in child component?

File: application/client/src/app/ui/views/workspace/headers/component.ts [hunk 5]
```diff
@@ -63,13 +69,44 @@ export class ColumnsHeaders extends ChangesDetector implements AfterContentInit
...
 
+    public ngOnMouseOver(header: string): void {
+        this.ngMouseOverHeader = header;
+        this.detectChanges();
+    }
+
+    public ngOnMouseLeave(): void {
+        this.ngMouseOverHeader = '';
+        this.detectChanges();
+    }
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/component.ts:93]: Tracking of mouse over and leave are quite expensive. Please, try to get rid of it. If you need `header` on event `click` you are able to get it with click event and no needs to track it during mouse movements.

Comment (DmitryAstafyev) [reply application/client/src/app/ui/views/workspace/headers/component.ts:93]: To show/hide `...` on header you can use `css`, with `::hover` 
```
& .header{
    & mat-icon {
       display: none;
    }
    &:hover {
    & mat-icon {
       display: block;
    }
    }
}
```

**Review by DmitryAstafyev — CHANGES_REQUESTED** [review]
General:
- styles are not okay: `...` icon isn't aligned vertically in the header; also an issue with aligning in the context menu on the list of headers;
- this context menu should be called by clicking on `...` and also with the context menu (right mouse click)
- no need to have there button `ok`, click outside, and `ESC` should close it as well

File: application/client/src/app/schema/render/columns.ts [hunk 6]
```diff
@@ -52,12 +59,23 @@ export class Columns {
                               widths[i],
                           ),
                 visible: headersVisability[i],
+                color: '#FFFFFF',
```
Comment (DmitryAstafyev) [inline application/client/src/app/schema/render/columns.ts:62]: please take color from `scheme` (`@ui/styles/colors.ts`)

File: application/client/src/app/ui/views/workspace/headers/component.ts [hunk 7]
```diff
@@ -63,13 +68,33 @@ export class ColumnsHeaders extends ChangesDetector implements AfterContentInit
...
+            this.controller.subjects.get().visibility.subscribe(() => {
+                this.headers = this.controller.headers
+                .filter((h) => h.visible)
+                .map((h) => new RenderedHeader(h));
+                this.markChangesForCheck();
+            })
         );
         this.headers = this.controller.headers
             .filter((h) => h.visible)
             .map((h) => new RenderedHeader(h));
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/component.ts:80]: I think here can be a problem. Property `headers` in `html` uses index of header to update its properties (`visibility` and `width`), but `index` can be changed... for example, you have at initianal state columns
```
column - index
a - 0
b - 1
c - 2
d - 3
e - 4
```
At some moment user hides column `b` and you will get
```
column - index
a - 0
c - 1
d - 2
e - 3
```
Indexes of columns c, d, and e are now changed...it's different in view and controller

File: application/client/src/app/ui/views/workspace/headers/menu/template.html [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<div class="wrapper">
+    <ul class="columns">
+        <li class="column"
+            *ngFor="let header of controller.headers; let i = index;"
+            (click)="ngOnCheckboxClick($event, i)">
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/menu/template.html:5]: looks like we need `uuid` of each column to correctly bind it between view and controller or static `index` (as property in header description)

File: application/client/src/app/ui/views/workspace/headers/menu/component.ts [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Input() public column!: number;
+    @Input() public controller!: Columns;
+
+    constructor(cdRef: ChangeDetectorRef) {
+        super(cdRef);
+    }
+
+    public ngOnCheckboxClick(event: MouseEvent, i: number): void {
+        event.stopPropagation();
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/menu/component.ts:29]: please use tools from `@ui/env/dom`... there is method `stop`

File: application/client/src/app/ui/views/workspace/headers/template.html [hunk 10]
```diff
@@ -1,10 +1,17 @@
 <span class="offset" [ngStyle]="ngGetOffsetStyle()"></span>
-<span class="header" 
+<span class="header"
       *ngFor="let header of headers; let i = index"
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/template.html:3]: looks like we need `uuid` of each column to correctly bind it between view and controller or static `index` (as property in header description)

File: application/client/src/app/ui/views/workspace/headers/styles.less [hunk 11]
```diff
@@ -22,10 +22,25 @@
...
+    }
     & span:nth-child(2) {
         border-left: none;
     }
+    // span.offset {
+    //     position: relative;
+    //     display: inline-block;
+    //     margin: 0;
+    //     padding: 0 0 0 1px;
+    // }
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/styles.less:43]: please, remove, if isn't needed

File: application/client/src/app/ui/views/workspace/headers/component.ts [hunk 12]
```diff
@@ -48,6 +52,7 @@ class RenderedHeader {
 export class ColumnsHeaders extends ChangesDetector implements AfterContentInit {
     public readonly Direction = Direction;
     public offset: number = 0;
+    public ngMore: string = 'more_horiz';
```
Comment (DmitryAstafyev) [inline application/client/src/app/ui/views/workspace/headers/component.ts:55]: no need to have it as a property, just put a value of it into `html` directly... that's not dynamic property

File: application/client/src/app/schema/render/columns.ts [hunk 13]
```diff
@@ -67,6 +85,22 @@ export class Columns {
         return this.headers[column].visible;
     }
 
+    public setColor(column: number, color: string): void {
+        if (this.headers[column] === undefined) {
+            throw new Error(`Maximum ${this.headers.length} present in the file and tried to set the color of column at #${column}`);
+        }
+        this.headers[column].color = (color === 'rgba(0,0,0,0)') ? '#FFF' : color;
```
Comment (DmitryAstafyev) [inline application/client/src/app/schema/render/columns.ts:92]: Please do not hardcode values of colors. Use theme/scheme from `@ui/styles/colors`

**Review by DmitryAstafyev — CHANGES_REQUESTED** [review]

**Comment by sudeeptarlekar** [comment]
@DmitryAstafyev resolved all resizing issues, pls review again.
Thanks in advance.

**Comment by DmitryAstafyev** [comment]
replaced with #1904

**Pull request closed without merging**